### PR TITLE
Issue2912 district cooling

### LINKED ISSUE #2912: Addition of district cooling model: ETS, Loads and Examples
Hi Dr. Hu @JayHuLBL , when you have a chance, could you please create a new branch for this issue so that I could add the developments?

This issue is to add the district cooling models to `Buildings.Experimental.DHC` package, including the followings:

### Buildings.Experimental.DHC
image
image
image

**Model Descriptions**
These models including three typical types of energy transfer station, one time series building loads, and an example model for the integral district cooling system. This work is part of the URBANopt project and will be extended for GMT(Geojson Modelica Translator) usage.

Our team still need to 2 weeks (targeted to March 21st) to complete the internal review, if you have any questions, please let me know. Thanks a lot!

### PR DESCRIPTION
**Pull request by JayHuLBL**
This closes #2912 and #2308.

### CODE REVIEW

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo
Comment (JayHuLBL) [inline Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo:182]: propagate the `bandwidth` to be an advanced parameter. Finalized the parameter assignment.

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo
Comment (JayHuLBL) [inline Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo:189]: at the end of the comments, should not have the period `.`. So the comment will be
`Full opening of the control valve`

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Indirect.mo
Comment (JayHuLBL) [inline Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Indirect.mo:152]: add comment

File: Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo
Comment (JayHuLBL) [inline Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo:131]: tab in

File: Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo
Comment (JayHuLBL) [inline Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo:174]: tab in

**Review by JayHuLBL — COMMENTED** [review]
@khinkelman Please see inline comments.

**Comment by JayHuLBL** [comment]
@khinkelman Please see the inline comments. I have also committed some changes [aada46d](https://github.com/lbl-srg/modelica-buildings/pull/3093/commits/aada46d5433c329b27b054d7015f4bdb3ff1f4fa), [9bd530d](https://github.com/lbl-srg/modelica-buildings/pull/3093/commits/9bd530d7be8f680b8097985229b0df9d9cc4554b), [4a87098](https://github.com/lbl-srg/modelica-buildings/pull/3093/commits/4a8709843697bef604e6d901c94dfd575e24345a), [551c8a3](https://github.com/lbl-srg/modelica-buildings/pull/3093/commits/551c8a3cc69a6c5eb5d4c68ce26a974f2601438e), to clean up models and reference results.
The model `Buildings.Experimental.DHC.Examples.Cooling.DirectUncontrolledETS` failed to run with Optimica and OpenModelica with following errors:

```
See 'simulator-optimica.log' for details.
409*** Error: Simulation of Buildings.Experimental.DHC.Examples.Cooling.DirectUncontrolledETS failed: 'FMUException: Failed to update the events at time: 1.572482E+07.'. Directory is '/tmp/tmp-Buildings-0-xlzir_zy'.
410Execution time = 1452.984 s
411make: *** [Makefile:150: test-optimica] Error 1
```

```
See 'simulator-openmodelica.log' for details.
385*** Error: Simulation of Buildings.Experimental.DHC.Examples.Cooling.DirectUncontrolledETS failed: 'CalledProcessError: Command '['omc', 'Buildings.Experimental.DHC.Examples.Cooling.DirectUncontrolledETS_simulate.mos']' returned non-zero exit status 1.'. Directory is '/tmp/tmp-Buildings-0-exrbuqkj'.
386Execution time = 1724.656 s
387make: *** [Makefile:145: test-openmodelica] Error 1
```

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo
Comment (khinkelman) [reply Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo:182]: Done

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo
Comment (khinkelman) [reply Buildings/Experimental/DHC/EnergyTransferStations/Cooling/DirectControlled.mo:189]: Done for this model.

- [ ] Check other models on this PR as well

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Indirect.mo
Comment (khinkelman) [reply Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Indirect.mo:152]: Done

File: Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo
Comment (khinkelman) [reply Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo:131]: Done

File: Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo
Comment (khinkelman) [reply Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo:174]: Done

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [inline Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:10]: Update filename to `DirectUncontrolledETS.mat`

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [inline Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:2]: - [ ] Fix simulation errors in this model
- [ ] Ensure all plot variables are appearing correctly

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [inline Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:14]: Address error:
``` plotExpression(apply(DistrictCooling[end].buiETS[1].bui.loa.y[1]+DistrictCooling[end].buiETS[2].bui.loa.y[1]+DistrictCooling[end].buiETS[3].bui.loa.y[1]), false, "loaTot [W]", 1);
Failed to expand plotExpression(apply(DistrictCooling[end].buiETS[1].bui.loa.y[1]+DistrictCooling  [end].buiETS[2].bui.loa.y[1]+DistrictCooling[end].buiETS[3].bui.loa.y[1]),   false, "loaTot [W]", 1, 1, "").
```

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Loads/Cooling/Examples/BuildingTimeSeriesWithETS.mos
Comment (khinkelman) [inline Buildings/Resources/Scripts/Dymola/Experimental/DHC/Loads/Cooling/Examples/BuildingTimeSeriesWithETS.mos:24]: - [ ] Remove duplicate variables
- [ ] Have supply and return temperatures plotted in the same frame

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [inline Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:1]: Address OCT Errors:
```
Warning: Possible chattering detected at t = 1.572480e+07 in state event(s): [4]
Traceback (most recent call last):
  File "/home/*/proj/06_runOCT/jmodelica.py", line 89, in <module>
    res = mod.simulate(options=opts)
  File "src/pyfmi/fmi.pyx", line 8125, in pyfmi.fmi.FMUModelME2.simulate
  File "src/pyfmi/fmi.pyx", line 378, in pyfmi.fmi.ModelBase._exec_simulate_algorithm
  File "src/pyfmi/fmi.pyx", line 374, in pyfmi.fmi.ModelBase._exec_simulate_algorithm
  File "install/Python/pyfmi/fmi_algorithm_drivers.py", line 691, in solve
  File "assimulo/ode.pyx", line 180, in assimulo.ode.ODE.simulate
  File "assimulo/ode.pyx", line 300, in assimulo.ode.ODE.simulate
  File "assimulo/explicit_ode.pyx", line 101, in assimulo.explicit_ode.Explicit_ODE._simulate
  File "assimulo/explicit_ode.pyx", line 225, in assimulo.explicit_ode.Explicit_ODE._simulate
  File "src/pyfmi/simulation/assimulo_interface.pyx", line 994, in pyfmi.simulation.assimulo_interface.FMIODE2.handle_event
  File "src/pyfmi/fmi.pyx", line 7862, in pyfmi.fmi.FMUModelME2.event_update
pyfmi.fmi.FMUException: Failed to update the events at time: 1.572480E+07.
```

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [reply Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:10]: Done

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [reply Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:2]: Done for Dymola

File: Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos
Comment (khinkelman) [reply Buildings/Resources/Scripts/Dymola/Experimental/DHC/Examples/Cooling/DirectUncontrolledETS.mos:14]: Removed these expressions.

**Comment by khinkelman** [comment]
Hi @JayHuLBL, I hope you are doing well. Just a friendly follow up on this PR and an update from my end. As of last Friday, I can now successfully run OCT on my workstation. 

Do you need anything from me to merge and close this contribution? Many thanks.

**Comment by JayHuLBL** [comment]
@khinkelman I rerun the unit test for the models in package `Buildings.Experimental.DHC.Plants.Cooling.Examples` and added the `Buildings.Experimental.DHC.Examples.Cooling.DirectUncontrolledETS` to the exclusion list to not run the model with `OpenModelica`. I have no further comments.

When the CI tests passed, it should be ready for Michael to review.

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Examples/DirectControlled.mo
Comment (mwetter) [inline Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Examples/DirectControlled.mo:2]: In this model, `cheVal.dp` has a few 100 bar pressure drop, possibly because the flow is forced through a closed 2 way valve. This needs to be corrected.

**Review by mwetter — CHANGES_REQUESTED** [review]
I am not yet done with the review, but here is a discrepancy that needs to be addressed.

File: Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Examples/DirectControlled.mo
Comment (mwetter) [inline Buildings/Experimental/DHC/EnergyTransferStations/Cooling/Examples/DirectControlled.mo:23]: That looks wrong. At the minimum, it should be some fraction of the design flow rate rather than just `0.01`.

**Comment by mwetter** [comment]
I will further process this after https://github.com/lbl-srg/modelica-buildings/pull/3157 is merged to the master (as https://github.com/lbl-srg/modelica-buildings/pull/3157 changes some of the underlying models).

File: Buildings/Experimental/DHC/Loads/Cooling/BuildingTimeSeriesWithETS.mo
Comment (mwetter) [inline Buildings/Experimental/DHC/Loads/Cooling/BuildingTimeSeriesWithETS.mo:15]: factor 40 and division by -1.5E40 is not clear. Document or remove.

**Comment by mwetter** [comment]
@khinkelman : This is ready to merge when the reference results are updated (which I can do) and when the review comment
https://github.com/lbl-srg/modelica-buildings/pull/3093#pullrequestreview-1248749396
is addressed. This goes back to a revision by @ChengnanShi-Work:
https://github.com/lbl-srg/modelica-buildings/blame/041de1b9ee39608da70bcaf3205c4ab24c77b7af/Buildings/Experimental/DHC/Loads/Cooling/BuildingTimeSeriesWithETS.mo#L8
Do you or @ChengnanShi-Work know what the rationale is? This should either be removed or explained if it is indeed needed.

**Comment by khinkelman** [comment]
> @khinkelman : This is ready to merge when the reference results are updated (which I can do) and when the review comment [#3093 (review)](https://github.com/lbl-srg/modelica-buildings/pull/3093#pullrequestreview-1248749396) is addressed. This goes back to a revision by @ChengnanShi-Work: https://github.com/lbl-srg/modelica-buildings/blame/041de1b9ee39608da70bcaf3205c4ab24c77b7af/Buildings/Experimental/DHC/Loads/Cooling/BuildingTimeSeriesWithETS.mo#L8 Do you or @ChengnanShi-Work know what the rationale is? This should either be removed or explained if it is indeed needed.

Thanks @mwetter for your review. I cannot recall the rational for scaling `facMulCoo` in that way. @AntoineGautier, could you please comment here? This setting of `facMulCoo=40*QCoo_flow_nominal/(-1.5E5))` came from your [Combined.BuildingTimeSeriesWithETS](https://github.com/lbl-srg/modelica-buildings/blob/041de1b9ee39608da70bcaf3205c4ab24c77b7af/Buildings/Experimental/DHC/Loads/Combined/BuildingTimeSeriesWithETS.mo) originally, from what I can tell.

**Comment by AntoineGautier** [comment]
These hard-coded numerical values should probably be removed. They come from an example model which was rapidly promoted as a generic class with lack of scrutiny regarding this parameter binding. The current PR #3165 provides a default assignment for `facMulCoo` that should rather be used. @dhblum is working on improving the documentation and the code itself which still includes some literal assignments. The rationale is the following.

- A default effectiveness is assigned based on fan coil manufacturer data: this gives the shape of the so-called emission characteristic of the terminal unit Q_flow = f(m_flow).
- The number of terminal units is computed based on the peak load and datasheet capacity.
- An adjustment is made to guard against modifications of air flow or entering temperatures by the user that would be inconsistent with the number of terminal units, i.e., yielding an effectiveness above 1: this is done by adjusting `facMul(Hea|Coo)` at fixed effectiveness. The effectiveness can be modified by manually overwriting the default assignment of `facMul(Hea|Coo)`.
- What is worth mentioning is that the parameterization always ensures that the specified water dT is achieved at design flows and entering temperatures.

**Comment by mwetter** [comment]
@dhblum Is it best to clean up and merge https://github.com/lbl-srg/modelica-buildings/pull/3165 and then process this PR? 
For this PR, I tried to remove the assignment to `facMulHea` and `facMulCoo`, but then `Buildings/Experimental/DHC/Loads/Cooling/Examples/BuildingTimeSeriesWithETS` won't initialize (as the cooling coil design parameters seem to be not meaningful).

**Comment by dhblum** [comment]
Yes, https://github.com/lbl-srg/modelica-buildings/pull/3165 should be merged first.  In that PR, the factors in question in https://github.com/lbl-srg/modelica-buildings/pull/3093#issuecomment-1384480665 are effectively moved to defaults within a base model (Buildings.Experimental.DHC.Loads.BaseClasses.Examples.BaseClasses.BuildingTimeSeries) so they won't have to be re-specified in this district cooling model.

File: Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo
Comment (mwetter) [inline Buildings/Experimental/DHC/Loads/Cooling/BaseClasses/BuildingTimeSeries.mo:119]: Why do we need two columns, one being for heating, if this model is only for cooling?

**Comment by mwetter** [comment]
If all tests pass, this is ready to be merged.

**Review by mwetter — APPROVED** [review]

**Comment by mwetter** [comment]
@dhblum : This is ready to merge if the unit tests succeed. Let me know if I should go ahead or whether you want to do another review.

File: Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo
Comment (dhblum) [inline Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo:154]: I recommend calling "domestic hot water load" instead of "sanitary" to be consistent with typical convention.  This is also used elsewhere in model.

File: Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo
Comment (dhblum) [inline Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo:387]: Suggest removing duplicate description of "Fourth column."

File: Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo
Comment (dhblum) [inline Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo:415]: Suggest using "domestic hot water load" rather than "sanitary."

File: Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo
Comment (dhblum) [inline Buildings/Experimental/DHC/Loads/BaseClasses/Examples/BaseClasses/BuildingTimeSeries.mo:18]: In Line 17, suggest changing SHW to "domestic hot water."

**Review by dhblum — CHANGES_REQUESTED** [review]
Thanks @mwetter for all the updates.  I think it is good to merge except for a few inline comments regarding the use of "sanitary" vs. "domestic hot" water.

**Comment by mwetter** [comment]
@dhblum : Thanks for the feedback, I addressed them in the latest commit and will merge when the tests passed.

**Comment by dhblum** [comment]
Sounds good, thanks @mwetter!

**Pull request merged**
